gdal/frmts/grib/gribdataset.cpp

```diff
@@ -824,11 +836,22 @@ CPLErr GRIBRasterBand::IReadBlock( int /* nBlockXOff */,
     if( nBlockYOff >= nGribDataYSize )  // Off image?
         return CE_None;
 
+    if ( nRasterXSize != nGribDataXSize && poGDS->nSplitAndSwapColumn != 0)
```

Comment:
could that check be moved to where nSplitAndSwapColumn is set ? Otherwise the georeferencing exposed will be inconsistant with the raster returned

Reply:
> The problem is that the GRIB is read sequentially so it could easily be the last rasterband that will trigger this condition

hum, that's a good point... anyway handing of such files is a bit broken

> do you happen to have a GRIB file with this?

9d629afc0b983fcec70a9c5a85c92db429ed4996 points to https://trac.osgeo.org/gdal/ticket/3246

Reply:
@rouault in fact, no matter what one does, the georeferencing of a band that doesn't match its dataset size will be incorrect
In the file from the old bug there is only one band that has a very different shape from the dataset (28x21 vs 103x78)
I wonder what you are supposed to do with this band - one can still retrieve the data (I will disable split&swap only for this band) but there won't be a valid geotransform anyway

Reply:
@rouault I think this dataset is corrupted, there is no other explanation
This is a sea swell grib and the weird band contains the U-wind, while the V-wind is in a normal bind
There is absolutely no reason for the U-wind to have a different shape from the V-wind
U and V are the weather science way of expressing the wind (separately on the NS and EW axis) instead of a speed and angle so that interpolation and resampling can work in a mathematically natural way
I have no explanation for this other that the dataset is corrupted

Reply:
especially since the values contained for the V wind look like wind, while U is total garbage

Reply:
beyond the brokenness of that particular example, we should ideally isolate in separate subdatasets GRIB messages that have different sizes/georeferencing, but that might be tricky to do in a performant way since currently we get the actual size in LoadData(). I guess that at least the size could be obtained in a cheaper way without decompressing the grid, but not sure we have a degrib API for that. Anyway this is a bit out of topic of this PR. One issue at a time :-)